# OLAP Cubes

## 1. Slicing and Dicing

All the databases table in this demo are based on public database samples and transformations
- `Sakila` is a sample database created by `MySql` [Link](https://dev.mysql.com/doc/sakila/en/sakila-structure.html)
- The postgresql version of it is called `Pagila` [Link](https://github.com/devrimgunduz/pagila)
- The facts and dimension tables design is based on O'Reilly's public dimensional modelling tutorial schema [Link](http://archive.oreilly.com/oreillyschool/courses/dba3/index.html)

Start by creating and connecting to the database

In [1]:
!PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila_star
!PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila_star -f Data/pagila-star.sql

 set_config 
------------
 
(1 row)

 setval 
--------
    200
(1 row)

 setval 
--------
    605
(1 row)

 setval 
--------
     16
(1 row)

 setval 
--------
    600
(1 row)

 setval 
--------
    109
(1 row)

 setval 
--------
    599
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
      1
(1 row)

 setval 
--------
  16049
(1 row)

 setval 
--------
   1000
(1 row)

 setval 
--------
   4581
(1 row)

 setval 
--------
      6
(1 row)

 setval 
--------
  32098
(1 row)

 setval 
--------
  16049
(1 row)

 setval 
--------
      2
(1 row)

 setval 
--------
      2
(1 row)



### Connect to the local database where Pagila is loaded

In [2]:
import sql
%load_ext sql

DB_ENDPOINT = "127.0.0.1"
DB = 'pagila_star'
DB_USER = 'student'
DB_PASSWORD = 'student'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)
%sql $conn_string

postgresql://student:student@127.0.0.1:5432/pagila_star


'Connected: student@pagila_star'

### Star Schema

<img src="pagila-star.png" width="50%"/>

### Starting with a simple cube
Calculate the revenue (sales_amount) by day, rating, and city. Then, sort by revenue in descending order and limit to the first 20 rows.

In [3]:
%%time
%%sql

SELECT d.day AS day, m.rating AS rating, c.city AS city, SUM(f.sales_amount) AS revenue
FROM factsales f 
JOIN dimDate d ON f.date_key = d.date_key
JOIN dimMovie m on f.movie_key = m.movie_key
JOIN dimCustomer c on f.customer_key = c.customer_key
GROUP BY day, rating, city
ORDER BY revenue DESC LIMIT 20

 * postgresql://student:***@127.0.0.1:5432/pagila_star
20 rows affected.
CPU times: user 4.93 ms, sys: 117 µs, total: 5.05 ms
Wall time: 41.8 ms


[(30, 'G', 'San Bernardino', Decimal('24.97')),
 (30, 'NC-17', 'Apeldoorn', Decimal('23.95')),
 (21, 'NC-17', 'Belm', Decimal('22.97')),
 (21, 'G', 'Citt del Vaticano', Decimal('21.97')),
 (28, 'R', 'Mwanza', Decimal('21.97')),
 (30, 'PG-13', 'Zanzibar', Decimal('21.97')),
 (22, 'R', 'Yangor', Decimal('19.97')),
 (28, 'PG-13', 'Dhaka', Decimal('19.97')),
 (1, 'R', 'Qomsheh', Decimal('19.97')),
 (17, 'G', 'Rajkot', Decimal('19.97')),
 (19, 'PG', 'Najafabad', Decimal('19.96')),
 (30, 'R', 'Fengshan', Decimal('19.95')),
 (28, 'PG', 'So Leopoldo', Decimal('18.98')),
 (21, 'G', 'Wroclaw', Decimal('18.98')),
 (19, 'PG', 'Sokoto', Decimal('18.97')),
 (30, 'G', 'Omdurman', Decimal('18.97')),
 (30, 'PG-13', 'Osmaniye', Decimal('18.97')),
 (1, 'NC-17', 'Memphis', Decimal('18.97')),
 (29, 'PG-13', 'Shimoga', Decimal('18.97')),
 (21, 'PG-13', 'Asuncin', Decimal('18.95'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>San Bernardino</td>
        <td>24.97</td>
    </tr>
    <tr>
        <td>30</td>
        <td>NC-17</td>
        <td>Apeldoorn</td>
        <td>23.95</td>
    </tr>
    <tr>
        <td>21</td>
        <td>NC-17</td>
        <td>Belm</td>
        <td>22.97</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Zanzibar</td>
        <td>21.97</td>
    </tr>
    <tr>
        <td>28</td>
        <td>R</td>
        <td>Mwanza</td>
        <td>21.97</td>
    </tr>
</tbody></table></div>

## Slicing

Slicing is the reduction of the dimensionality of a cube by 1 e.g. 3 dimensions to 2, fixing one of the dimensions to a single value. In the example above, we have a 3-dimensional cube on day, rating, and country.

Reduce the dimensionality of the above example by limiting the results to only include movies with a `rating` of "PG-13". Again, sort by revenue in descending order and limit to the first 20 rows.

In [4]:
%%time
%%sql

SELECT d.day AS day, m.rating AS rating, c.city AS city, SUM(f.sales_amount) AS revenue
FROM factsales f 
JOIN dimDate d ON f.date_key = d.date_key
JOIN dimMovie m on f.movie_key = m.movie_key
JOIN dimCustomer c on f.customer_key = c.customer_key
WHERE m.rating='PG-13'
GROUP BY day, rating, city
ORDER BY revenue DESC LIMIT 20

 * postgresql://student:***@127.0.0.1:5432/pagila_star
20 rows affected.
CPU times: user 4.77 ms, sys: 94 µs, total: 4.86 ms
Wall time: 20.3 ms


[(30, 'PG-13', 'Zanzibar', Decimal('21.97')),
 (28, 'PG-13', 'Dhaka', Decimal('19.97')),
 (30, 'PG-13', 'Osmaniye', Decimal('18.97')),
 (29, 'PG-13', 'Shimoga', Decimal('18.97')),
 (21, 'PG-13', 'Asuncin', Decimal('18.95')),
 (20, 'PG-13', 'Baha Blanca', Decimal('17.98')),
 (30, 'PG-13', 'Nagareyama', Decimal('17.98')),
 (21, 'PG-13', 'Parbhani', Decimal('17.98')),
 (30, 'PG-13', 'Tanauan', Decimal('17.96')),
 (17, 'PG-13', 'Ikerre', Decimal('17.95')),
 (30, 'PG-13', 'Zhoushan', Decimal('16.98')),
 (30, 'PG-13', 'Yerevan', Decimal('16.97')),
 (30, 'PG-13', 'Newcastle', Decimal('16.97')),
 (8, 'PG-13', 'Pontianak', Decimal('15.98')),
 (17, 'PG-13', 'Yantai', Decimal('15.98')),
 (30, 'PG-13', 'Santa Rosa', Decimal('15.98')),
 (20, 'PG-13', 'Santa Rosa', Decimal('15.98')),
 (30, 'PG-13', 'Toulouse', Decimal('15.97')),
 (23, 'PG-13', 'Tiefa', Decimal('15.97')),
 (10, 'PG-13', 'Udaipur', Decimal('15.97'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Zanzibar</td>
        <td>21.97</td>
    </tr>
    <tr>
        <td>28</td>
        <td>PG-13</td>
        <td>Dhaka</td>
        <td>19.97</td>
    </tr>
    <tr>
        <td>29</td>
        <td>PG-13</td>
        <td>Shimoga</td>
        <td>18.97</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Osmaniye</td>
        <td>18.97</td>
    </tr>
    <tr>
        <td>21</td>
        <td>PG-13</td>
        <td>Asuncin</td>
        <td>18.95</td>
    </tr>
</tbody></table></div>

## Dicing
Dicing is creating a subcube with the same dimensionality but fewer values for  two or more dimensions. 

Create a subcube of the initial cube that includes movies with:
* ratings of PG or PG-13
* in the city of Bellevue or Lancaster
* day equal to 1, 15, or 30

In [5]:
%%time
%%sql

SELECT d.day AS day, m.rating AS rating, c.city AS city, SUM(f.sales_amount) AS revenue
FROM factsales f 
JOIN dimDate d ON f.date_key = d.date_key
JOIN dimMovie m on f.movie_key = m.movie_key
JOIN dimCustomer c on f.customer_key = c.customer_key
WHERE m.rating IN ('PG-13', 'PG') AND c.city IN ('Bellevue', 'Lancaster') AND d.day IN (1,15,30)
GROUP BY day, rating, city
ORDER BY revenue DESC LIMIT 20

 * postgresql://student:***@127.0.0.1:5432/pagila_star
6 rows affected.
CPU times: user 5.33 ms, sys: 167 µs, total: 5.49 ms
Wall time: 10.2 ms


[(30, 'PG', 'Lancaster', Decimal('12.98')),
 (1, 'PG-13', 'Lancaster', Decimal('5.99')),
 (30, 'PG-13', 'Bellevue', Decimal('3.99')),
 (30, 'PG-13', 'Lancaster', Decimal('2.99')),
 (15, 'PG-13', 'Bellevue', Decimal('1.98')),
 (1, 'PG', 'Bellevue', Decimal('0.99'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>PG</td>
        <td>Lancaster</td>
        <td>12.98</td>
    </tr>
    <tr>
        <td>1</td>
        <td>PG-13</td>
        <td>Lancaster</td>
        <td>5.99</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Bellevue</td>
        <td>3.99</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Lancaster</td>
        <td>2.99</td>
    </tr>
    <tr>
        <td>15</td>
        <td>PG-13</td>
        <td>Bellevue</td>
        <td>1.98</td>
    </tr>
</tbody></table></div>

## 2. Roll-ups and Drill-downs

### Roll-up
- Stepping up the level of aggregation to a large grouping
- e.g.`city` is summed as `country`

Calculate revenue (sales_amount) by day, rating, and country. Sort the data by revenue in descending order, and limit the data to the top 20 results.

In [6]:
%%time
%%sql
SELECT d.day AS day, m.rating AS rating, c.country AS country, SUM(f.sales_amount) AS revenue
FROM factsales f 
JOIN dimDate d ON f.date_key = d.date_key
JOIN dimMovie m on f.movie_key = m.movie_key
JOIN dimCustomer c on f.customer_key = c.customer_key
GROUP BY day, rating, country
ORDER BY revenue DESC LIMIT 20

 * postgresql://student:***@127.0.0.1:5432/pagila_star
20 rows affected.
CPU times: user 4.93 ms, sys: 0 ns, total: 4.93 ms
Wall time: 30.5 ms


[(30, 'G', 'China', Decimal('169.67')),
 (30, 'PG', 'India', Decimal('156.67')),
 (30, 'NC-17', 'India', Decimal('153.64')),
 (30, 'PG-13', 'China', Decimal('146.67')),
 (30, 'R', 'China', Decimal('145.66')),
 (30, 'R', 'India', Decimal('143.68')),
 (30, 'G', 'India', Decimal('137.67')),
 (18, 'NC-17', 'India', Decimal('135.75')),
 (30, 'PG', 'China', Decimal('131.72')),
 (21, 'PG-13', 'India', Decimal('128.74')),
 (18, 'PG-13', 'India', Decimal('121.72')),
 (18, 'PG', 'India', Decimal('119.76')),
 (30, 'PG-13', 'India', Decimal('116.72')),
 (21, 'NC-17', 'China', Decimal('115.77')),
 (21, 'R', 'India', Decimal('115.75')),
 (27, 'NC-17', 'India', Decimal('115.72')),
 (17, 'PG-13', 'China', Decimal('111.75')),
 (17, 'NC-17', 'United States', Decimal('109.76')),
 (30, 'NC-17', 'China', Decimal('108.77')),
 (20, 'PG-13', 'India', Decimal('101.79'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>China</td>
        <td>169.67</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG</td>
        <td>India</td>
        <td>156.67</td>
    </tr>
    <tr>
        <td>30</td>
        <td>NC-17</td>
        <td>India</td>
        <td>153.64</td>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>China</td>
        <td>146.67</td>
    </tr>
    <tr>
        <td>30</td>
        <td>R</td>
        <td>China</td>
        <td>145.66</td>
    </tr>
</tbody></table></div>

### Drill-down
- Breaking up one of the dimensions to a lower level.
- e.g.`city` is broken up into  `districts`

Calculate revenue (sales_amount) by day, rating, and district. Sort the data by revenue in descending order, and limit the data to the top 20 results.

In [7]:
%%time
%%sql
SELECT d.day AS day, m.rating AS rating, c.district AS district, SUM(f.sales_amount) AS revenue
FROM factsales f 
JOIN dimDate d ON f.date_key = d.date_key
JOIN dimMovie m on f.movie_key = m.movie_key
JOIN dimCustomer c on f.customer_key = c.customer_key
GROUP BY day, rating, district
ORDER BY revenue DESC LIMIT 20

 * postgresql://student:***@127.0.0.1:5432/pagila_star
20 rows affected.
CPU times: user 2.09 ms, sys: 3.49 ms, total: 5.57 ms
Wall time: 38.1 ms


[(30, 'PG-13', 'Southern Tagalog', Decimal('53.88')),
 (30, 'G', 'Inner Mongolia', Decimal('38.93')),
 (30, 'G', 'Shandong', Decimal('36.93')),
 (30, 'NC-17', 'West Bengali', Decimal('36.92')),
 (17, 'PG-13', 'Shandong', Decimal('34.95')),
 (1, 'PG', 'California', Decimal('32.94')),
 (18, 'NC-17', 'So Paulo', Decimal('32.93')),
 (30, 'NC-17', 'Buenos Aires', Decimal('31.93')),
 (21, 'R', 'So Paulo', Decimal('31.93')),
 (30, 'PG', 'Southern Tagalog', Decimal('30.94')),
 (30, 'PG', 'So Paulo', Decimal('30.93')),
 (30, 'R', 'Buenos Aires', Decimal('30.92')),
 (18, 'NC-17', 'Maharashtra', Decimal('29.95')),
 (30, 'G', 'California', Decimal('29.95')),
 (21, 'PG-13', 'Uttar Pradesh', Decimal('29.94')),
 (20, 'PG', 'Shandong', Decimal('29.93')),
 (10, 'R', 'Maharashtra', Decimal('29.93')),
 (18, 'R', 'Sumy', Decimal('28.97')),
 (21, 'PG-13', 'Southern Tagalog', Decimal('28.96')),
 (29, 'NC-17', 'So Paulo', Decimal('28.95'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>day</th>
        <th>rating</th>
        <th>district</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>30</td>
        <td>PG-13</td>
        <td>Southern Tagalog</td>
        <td>53.88</td>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>Inner Mongolia</td>
        <td>38.93</td>
    </tr>
    <tr>
        <td>30</td>
        <td>G</td>
        <td>Shandong</td>
        <td>36.93</td>
    </tr>
    <tr>
        <td>30</td>
        <td>NC-17</td>
        <td>West Bengali</td>
        <td>36.92</td>
    </tr>
    <tr>
        <td>17</td>
        <td>PG-13</td>
        <td>Shandong</td>
        <td>34.95</td>
    </tr>
</tbody></table></div>

## 3. Grouping Sets
- It happens often that for 3 dimensions, you want to aggregate a fact:
    - by nothing (total)
    - then by the 1st dimension
    - then by the 2nd 
    - then by the 3rd 
    - then by the 1st and 2nd
    - then by the 2nd and 3rd
    - then by the 1st and 3rd
    - then by the 1st and 2nd and 3rd
    
- Since this is very common, and in all cases, we are iterating through all the fact table anyhow, there is a more clever way to do that using the SQL grouping statement "GROUPING SETS" 

### Revenue Total, by Month, by Country, by Month & Country All in one shot

The query calculates total revenue at the various grouping levels (total, by month, by country, by month & country) all at once using the grouping sets function.

In [8]:
%%sql
SELECT d.month AS month, s.country AS country, SUM(sales_amount) AS revenue
FROM factsales f
JOIN dimDate d ON f.date_key = d.date_key
JOIN dimStore s ON f.store_key = s.store_key
GROUP BY GROUPING SETS ((), (month), (country), (month, country))

 * postgresql://student:***@127.0.0.1:5432/pagila_star
18 rows affected.


[(1, 'Australia', Decimal('2364.19')),
 (1, 'Canada', Decimal('2460.24')),
 (1, None, Decimal('4824.43')),
 (2, 'Australia', Decimal('4895.10')),
 (2, 'Canada', Decimal('4736.78')),
 (2, None, Decimal('9631.88')),
 (3, 'Australia', Decimal('12060.33')),
 (3, 'Canada', Decimal('11826.23')),
 (3, None, Decimal('23886.56')),
 (4, 'Australia', Decimal('14136.07')),
 (4, 'Canada', Decimal('14423.39')),
 (4, None, Decimal('28559.46')),
 (5, 'Australia', Decimal('271.08')),
 (5, 'Canada', Decimal('243.10')),
 (5, None, Decimal('514.18')),
 (None, None, Decimal('67416.51')),
 (None, 'Australia', Decimal('33726.77')),
 (None, 'Canada', Decimal('33689.74'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>month</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>1</td>
        <td>Australia</td>
        <td>2364.19</td>
    </tr>
    <tr>
        <td>1</td>
        <td>Canada</td>
        <td>2460.24</td>
    </tr>
    <tr>
        <td>1</td>
        <td>None</td>
        <td>4824.43</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Australia</td>
        <td>4895.10</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Canada</td>
        <td>4736.78</td>
    </tr>
    <tr>
        <td>2</td>
        <td>None</td>
        <td>9631.88</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Australia</td>
        <td>12060.33</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Canada</td>
        <td>11826.23</td>
    </tr>
    <tr>
        <td>3</td>
        <td>None</td>
        <td>23886.56</td>
    </tr>
    <tr>
        <td>4</td>
        <td>Australia</td>
        <td>14136.07</td>
    </tr>
    <tr>
        <td>4</td>
        <td>Canada</td>
        <td>14423.39</td>
    </tr>
    <tr>
        <td>4</td>
        <td>None</td>
        <td>28559.46</td>
    </tr>
    <tr>
        <td>5</td>
        <td>Australia</td>
        <td>271.08</td>
    </tr>
    <tr>
        <td>5</td>
        <td>Canada</td>
        <td>243.10</td>
    </tr>
    <tr>
        <td>5</td>
        <td>None</td>
        <td>514.18</td>
    </tr>
    <tr>
        <td>None</td>
        <td>None</td>
        <td>67416.51</td>
    </tr>
    <tr>
        <td>None</td>
        <td>Australia</td>
        <td>33726.77</td>
    </tr>
    <tr>
        <td>None</td>
        <td>Canada</td>
        <td>33689.74</td>
    </tr>
</tbody></table></div>

## 4. CUBE Queries
- Group by CUBE (dim1, dim2, ..) , produces all combinations of different lengths in one go.
- This view could be materialized in a view and queried which would save lots repetitive aggregations

Calculate the various levels of aggregation done in the grouping sets exercise (total, by month, by country, by month & country) using the CUBE function.

In [9]:
%%time
%%sql
SELECT d.month AS month, s.country AS country, SUM(sales_amount) AS revenue
FROM factsales f
JOIN dimDate d ON f.date_key = d.date_key
JOIN dimStore s ON f.store_key = s.store_key
GROUP BY CUBE (month, country)

 * postgresql://student:***@127.0.0.1:5432/pagila_star
18 rows affected.
CPU times: user 4.9 ms, sys: 0 ns, total: 4.9 ms
Wall time: 36.7 ms


[(1, 'Australia', Decimal('2364.19')),
 (1, 'Canada', Decimal('2460.24')),
 (1, None, Decimal('4824.43')),
 (2, 'Australia', Decimal('4895.10')),
 (2, 'Canada', Decimal('4736.78')),
 (2, None, Decimal('9631.88')),
 (3, 'Australia', Decimal('12060.33')),
 (3, 'Canada', Decimal('11826.23')),
 (3, None, Decimal('23886.56')),
 (4, 'Australia', Decimal('14136.07')),
 (4, 'Canada', Decimal('14423.39')),
 (4, None, Decimal('28559.46')),
 (5, 'Australia', Decimal('271.08')),
 (5, 'Canada', Decimal('243.10')),
 (5, None, Decimal('514.18')),
 (None, None, Decimal('67416.51')),
 (None, 'Australia', Decimal('33726.77')),
 (None, 'Canada', Decimal('33689.74'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>month</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>1</td>
        <td>Australia</td>
        <td>2364.19</td>
    </tr>
    <tr>
        <td>1</td>
        <td>Canada</td>
        <td>2460.24</td>
    </tr>
    <tr>
        <td>1</td>
        <td>None</td>
        <td>4824.43</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Australia</td>
        <td>4895.10</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Canada</td>
        <td>4736.78</td>
    </tr>
    <tr>
        <td>2</td>
        <td>None</td>
        <td>9631.88</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Australia</td>
        <td>12060.33</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Canada</td>
        <td>11826.23</td>
    </tr>
    <tr>
        <td>3</td>
        <td>None</td>
        <td>23886.56</td>
    </tr>
    <tr>
        <td>4</td>
        <td>Australia</td>
        <td>14136.07</td>
    </tr>
    <tr>
        <td>4</td>
        <td>Canada</td>
        <td>14423.39</td>
    </tr>
    <tr>
        <td>4</td>
        <td>None</td>
        <td>28559.46</td>
    </tr>
    <tr>
        <td>5</td>
        <td>Australia</td>
        <td>271.08</td>
    </tr>
    <tr>
        <td>5</td>
        <td>Canada</td>
        <td>243.10</td>
    </tr>
    <tr>
        <td>5</td>
        <td>None</td>
        <td>514.18</td>
    </tr>
    <tr>
        <td>None</td>
        <td>None</td>
        <td>67416.51</td>
    </tr>
    <tr>
        <td>None</td>
        <td>Australia</td>
        <td>33726.77</td>
    </tr>
    <tr>
        <td>None</td>
        <td>Canada</td>
        <td>33689.74</td>
    </tr>
</tbody></table></div>